Src/VimCore/CommonOperations.fs

```diff
@@ -2387,19 +2427,385 @@ type internal CommonOperations
     /// Map the specified point with the specified tracking to the current snapshot
     member x.MapPointToCurrentSnapshot (point: SnapshotPoint) (pointTrackingMode: PointTrackingMode) =
         let snapshot = _textBuffer.CurrentSnapshot
-        if point.Snapshot = snapshot then
-            point
+        TrackingPointUtil.GetPointInSnapshot point pointTrackingMode snapshot
+        |> Option.defaultValue (SnapshotPoint(snapshot, min point.Position snapshot.Length))
+
+    /// Map the specified caret point to the current snapshot
+    member x.MapCaretPointToCurrentSnapshot (point: VirtualSnapshotPoint) =
+        if _vimTextBuffer.UseVirtualSpace then
+            let snapshot = _textBuffer.CurrentSnapshot
+            let pointTrackingMode = PointTrackingMode.Negative
+            match TrackingPointUtil.GetVirtualPointInSnapshot point pointTrackingMode snapshot with
+            | Some point -> point
+            | None ->
+                let defaultPoint =
+                    SnapshotPoint(snapshot, min point.Position.Position snapshot.Length)
+                VirtualSnapshotPoint(defaultPoint, point.VirtualSpaces)
+        else
+            point.Position
+            |> x.MapPointNegativeToCurrentSnapshot
+            |> VirtualSnapshotPointUtil.OfPoint
+
+    /// Map the specified selected span to the current snapshot
+    member x.MapSelectedSpanToCurrentSnapshot (span: SelectedSpan) =
+        let caretPoint = x.MapCaretPointToCurrentSnapshot span.CaretPoint
+        let anchorPoint = x.MapCaretPointToCurrentSnapshot span.AnchorPoint
+        let activcePoint = x.MapCaretPointToCurrentSnapshot span.ActivePoint
+        SelectedSpan(caretPoint, anchorPoint, activcePoint)
+
+    /// Add a new caret at the specified point
+    member x.AddCaretAtPoint (point: VirtualSnapshotPoint) =
+        let contains = VirtualSnapshotSpanUtil.ContainsOrEndsWith
+        let isContainedByExistingSpan =
+            x.SelectedSpans
+            |> Seq.exists (fun span -> contains span.Span point)
+        let remainingSpans =
+            x.SelectedSpans
+            |> Seq.filter (fun span -> not (contains span.Span point))
+            |> Seq.toArray
+        if
+            remainingSpans.Length > 0
+            || not isContainedByExistingSpan
+        then
+            seq {
+                yield! remainingSpans
+                if not isContainedByExistingSpan then
+                    yield SelectedSpan(point)
+            }
+            |> x.SetSelectedSpans
+
+    /// Add the specified selected span
+    member x.AddSelectedSpan selectedSpan =

```

Comment:
Why does `AddSelectedSpan` not check for an existing span that overlaps the new one but `AddCaretPoint` does check for overlap? Would naiively expect the overlap behavior to be consistent between the two functions.

Reply:
Sorry, `AddCaretAtPoint` should be called `AddOrRemoveCaretAtPoint`. If you add a caret and then click it again, you remove that caret. By contrast, `AddSelectedSpan` is a utility function whose purpose is to always add a new selection." In fact, it is only used once, so it could be a local function.